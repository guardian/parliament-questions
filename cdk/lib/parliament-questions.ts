import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuS3Bucket } from '@guardian/cdk/lib/constructs/s3';
import { GuardianAwsAccounts } from '@guardian/private-infrastructure-config';
import { Duration } from 'aws-cdk-lib';
import type { App } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

const APP_NAME = 'parliament-questions';

export enum LoggingLevel {
	SILLY,
	TRACE,
	DEBUG,
	INFO,
	WARN,
	ERROR,
	FATAL,
}

export class ParliamentQuestions extends GuStack {
	constructor(
		scope: App,
		id: string,
		props: GuStackProps,
		loggingLevel: number = LoggingLevel.WARN,
	) {
		super(scope, id, props);

		const ssmPrefix = `arn:aws:ssm:${props.env?.region}:${GuardianAwsAccounts.Investigations}:parameter`;
		const ssmPath = `${this.stage}/${this.stack}/${APP_NAME}`;

		const parliamentWrittenQuestions = new GuS3Bucket(
			this,
			'ParliamentWrittenQuestionsBucket',
			{
				app: APP_NAME,
				bucketName: `parliament-written-questions-${this.stage.toLowerCase()}`,
			},
		);

		// we only want one dev bucket so only create on CODE
		if (props.stage === 'CODE') {
			new GuS3Bucket(
				this,
				'ParliamentWrittenQuestionsBucket',
				{
					app: APP_NAME,
					bucketName: `parliament-written-questions-dev`,
				},
			);
		}

		const otherConfig = {
			app: APP_NAME,
			runtime: Runtime.NODEJS_18_X,
			fileName: `${APP_NAME}.zip`,
			timeout: Duration.millis(45000),
			environment: {
				Bucket: `${APP_NAME}-dist`,
				Stage: this.stage,
				LoggingLevel: loggingLevel.toString(),
			},
		};

		const lambda = new GuScheduledLambda(this, APP_NAME, {
			handler: 'dist/lambda/index.handler',
			rules: [
				{
					schedule: Schedule.cron({ hour: '10', minute: '00', weekDay: '2' }),
				},
			],
			monitoringConfiguration: {
				noMonitoring: true,
			},
			...otherConfig,
		});

		const getParametersPolicy = new PolicyStatement({
			effect: Effect.ALLOW,
			actions: ['ssm:GetParameter', 'ssm:GetParametersByPath'],
			resources: [`${ssmPrefix}/${ssmPath}/*`],
		});

		lambda.addToRolePolicy(getParametersPolicy); 

		lambda.addToRolePolicy(
			new PolicyStatement({
				effect: Effect.ALLOW,
				actions: ['s3:PutObject'],
				resources: [`${parliamentWrittenQuestions.bucketArn}/*`],
			}),
		);
	}
}
