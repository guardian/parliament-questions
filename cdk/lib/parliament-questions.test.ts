import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ParliamentQuestions } from './parliament-questions';

describe('The ParliamentQuestions stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new ParliamentQuestions(app, 'ParliamentQuestions', {
			stack: 'investigations',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
