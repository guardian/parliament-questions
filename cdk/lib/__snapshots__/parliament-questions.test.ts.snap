// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ParliamentQuestions stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuS3Bucket",
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ParliamentWrittenQuestionsBucketParliamentquestions44C9D562": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "parliament-written-questions-test",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "parliament-questions",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/parliament-questions",
          },
          {
            "Key": "Stack",
            "Value": "investigations",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "parliamentquestions48FEC04C": {
      "DependsOn": [
        "parliamentquestionsServiceRoleDefaultPolicy15A0ED1E",
        "parliamentquestionsServiceRole2F139CA6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "investigations/TEST/parliament-questions/parliament-questions.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "parliament-questions",
            "Bucket": "parliament-questions-dist",
            "LoggingLevel": "4",
            "STACK": "investigations",
            "STAGE": "TEST",
            "Stage": "TEST",
          },
        },
        "Handler": "dist/lambda/index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "parliamentquestionsServiceRole2F139CA6",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "parliament-questions",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/parliament-questions",
          },
          {
            "Key": "Stack",
            "Value": "investigations",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 45,
      },
      "Type": "AWS::Lambda::Function",
    },
    "parliamentquestionsServiceRole2F139CA6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "parliament-questions",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/parliament-questions",
          },
          {
            "Key": "Stack",
            "Value": "investigations",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "parliamentquestionsServiceRoleDefaultPolicy15A0ED1E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/investigations/TEST/parliament-questions/parliament-questions.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/investigations/parliament-questions",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/investigations/parliament-questions/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameter",
                "ssm:GetParametersByPath",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:undefined:745349931791:parameter/TEST/investigations/parliament-questions/*",
            },
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "ParliamentWrittenQuestionsBucketParliamentquestions44C9D562",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": "s3:ListBucket",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ParliamentWrittenQuestionsBucketParliamentquestions44C9D562",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parliamentquestionsServiceRoleDefaultPolicy15A0ED1E",
        "Roles": [
          {
            "Ref": "parliamentquestionsServiceRole2F139CA6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "parliamentquestionsparliamentquestionscron00100AllowEventRuleParliamentQuestionsparliamentquestions179D0339095CAD34": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "parliamentquestions48FEC04C",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "parliamentquestionsparliamentquestionscron00100DB99E758",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "parliamentquestionsparliamentquestionscron00100DB99E758": {
      "Properties": {
        "ScheduleExpression": "cron(00 10 * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "parliamentquestions48FEC04C",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
